File: Leetcode-one-by-one/src/Minimum Absolute Difference in Sliding Submatrix.swift
import Foundation

class Solution {
    func minAbsDiff(_ grid: [[Int]], _ k: Int) -> [[Int]] {
        let m = grid.count
        let n = grid[0].count
        
        // Edge case
        if k == 1 {
            return Array(repeating: Array(repeating: 0, count: n), count: m)
        }
        
        var ans = Array(
            repeating: Array(repeating: 0, count: n - k + 1),
            count: m - k + 1
        )
        
        for i in 0...(m - k) {
            for j in 0...(n - k) {
                
                var arr: [Int] = []
                
                // Collect elements
                for x in i..<(i + k) {
                    for y in j..<(j + k) {
                        arr.append(grid[x][y])
                    }
                }
                
                arr.sort()
                
                var minDiff = Int.max
                
                for t in 1..<arr.count {
                    if arr[t] == arr[t - 1] { continue }
                    minDiff = min(minDiff, arr[t] - arr[t - 1])
                }
                
                // If all elements same
                if minDiff == Int.max {
                    minDiff = 0
                }
                
                ans[i][j] = minDiff
            }
        }
        
        return ans
    }
}

File: Leetcode-one-by-one/src/Minimum Absolute Difference in Sliding Submatrix.go
import (
    "math"
    "sort"
)

func minAbsDiff(grid [][]int, k int) [][]int {
    m, n := len(grid), len(grid[0])

    // Edge case
    if k == 1 {
        ans := make([][]int, m)
        for i := range ans {
            ans[i] = make([]int, n)
        }
        return ans
    }

    ans := make([][]int, m-k+1)
    for i := range ans {
        ans[i] = make([]int, n-k+1)
    }

    for i := 0; i <= m-k; i++ {
        for j := 0; j <= n-k; j++ {

            arr := make([]int, 0, k*k)

            // Collect elements
            for x := i; x < i+k; x++ {
                for y := j; y < j+k; y++ {
                    arr = append(arr, grid[x][y])
                }
            }

            sort.Ints(arr)

            minDiff := math.MaxInt32

            for t := 1; t < len(arr); t++ {
                if arr[t] == arr[t-1] {
                    continue
                }
                diff := arr[t] - arr[t-1]
                if diff < minDiff {
                    minDiff = diff
                }
            }

            // If all elements same
            if minDiff == math.MaxInt32 {
                minDiff = 0
            }

            ans[i][j] = minDiff
        }
    }

    return ans
}
